pkg/apis/extensions/v1alpha1/types_bastion.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var _ Object = (*Bastion)(nil)
+
+// BastionResource is a constant for the name of the Bastion resource.
+const BastionResource = "Bastion"
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Namespaced,path=bastions,singular=bastion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name=IP,JSONPath=".spec.bastionIP",type=string,description="The public IP address of the temporary bastion host"
+// +kubebuilder:printcolumn:name=State,JSONPath=".status.state",type=string,description="The bastion's status."
+// +kubebuilder:printcolumn:name=Age,JSONPath=".metadata.creationTimestamp",type=date,description="creation timestamp"
+
+// Bastion is a bastion or jump host that is dynamically created
+// to provide SSH access to shoot nodes.
+type Bastion struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Spec is the specification of this OpenIDConnect preset.
+	Spec BastionSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+	// Status is the bastion's status.
+	// +optional
+	Status BastionStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// GetExtensionSpec implements Object.
+func (b *Bastion) GetExtensionSpec() Spec {
+	return &b.Spec
+}
+
+// GetExtensionStatus implements Object.
+func (b *Bastion) GetExtensionStatus() Status {
+	return &b.Status
+}
+
+// BastionSpec contains the specification for an SSH bastion host.
+type BastionSpec struct {
+	// DefaultSpec is a structure containing common fields used by all extension resources.
+	DefaultSpec `json:",inline"`
+	// ShootRef is the target cluster to which a tunnel should be
+	// established.
+	ShootRef corev1.LocalObjectReference `json:"shootRef" protobuf:"bytes,1,name=shootRef"`
```

Comment:
OK, now I'm confused, I was assuming this is the `Bastion` resource in the `extensions.gardener.cloud` API group "living" in the seed cluster. The `Shoot` resources are only created in the garden cluster, so a `shootRef` doesn't make sense here?

Reply:
Correct, I started with the operations API group and then switched to the extensions (i.e. buiilding bottom-up instead of top-down), so I had some fields mixed up.